pkg/k8shandler/kibana/reconciler.go

```diff
@@ -0,0 +1,698 @@
+	clusterKibanaRequest := ClusterKibanaRequest{
+		client:  requestClient,
+		cluster: requestCluster,
+	}
+
+	if clusterKibanaRequest.cluster == nil {
+		return nil
+	}
+
+	if err := clusterKibanaRequest.CreateOrUpdateServiceAccount(kibanaServiceAccountName, &kibanaServiceAccountAnnotations); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.RemoveClusterRoleBinding("kibana-proxy-oauth-delegator"); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaService(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaRoute(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaConsoleExternalLogLink(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.RemoveOAuthClient("kibana-proxy"); err != nil {
+		return err
+	}
+
+	//proxyConfig := &configv1.Proxy{}
+	if err := clusterKibanaRequest.createOrUpdateKibanaDeployment(proxyConfig); err != nil {
+		return err
+	}
+
+	kibanaStatus, err := clusterKibanaRequest.getKibanaStatus()
+	cluster := clusterKibanaRequest.cluster
+
+	if err != nil {
+		return fmt.Errorf("Failed to get Kibana status for %q: %v", cluster.Name, err)
+	}
+
+	printUpdateMessage := true
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry,
+		func() error {
+			if !compareKibanaStatus(kibanaStatus,
+				cluster.Status) {
+				if printUpdateMessage {
+					logrus.Infof("Updating status of Kibana")
+					printUpdateMessage = false
+				}
+				cluster.Status = kibanaStatus
+				return clusterKibanaRequest.UpdateStatus(cluster)
+			}
+			return nil
+		})
+	if retryErr != nil {
+		return fmt.Errorf("Failed to update Kibana status for %q: %v", cluster.Name, retryErr)
+	}
+
+	return nil
+}
+
+func compareKibanaStatus(lhs, rhs []kibana.KibanaStatus) bool {
+	// there should only ever be a single kibana status object
+	if len(lhs) != len(rhs) {
+		return false
+	}
+
+	if len(lhs) > 0 {
+		for index, _ := range lhs {
+			if lhs[index].Deployment != rhs[index].Deployment {
+				return false
+			}
+
+			if lhs[index].Replicas != rhs[index].Replicas {
+				return false
+			}
+
+			if len(lhs[index].ReplicaSets) != len(rhs[index].ReplicaSets) {
+				return false
+			}
+
+			if len(lhs[index].ReplicaSets) > 0 {
+				if !reflect.DeepEqual(lhs[index].ReplicaSets, rhs[index].ReplicaSets) {
+					return false
+				}
+			}
+
+			if len(lhs[index].Pods) != len(rhs[index].Pods) {
+				return false
+			}
+
+			if len(lhs[index].Pods) > 0 {
+				if !reflect.DeepEqual(lhs[index].Pods, rhs[index].Pods) {
+					return false
+				}
+			}
+
+			if len(lhs[index].Conditions) != len(rhs[index].Conditions) {
+				return false
+			}
+
+			if len(lhs[index].Conditions) > 0 {
+				if !reflect.DeepEqual(lhs[index].Conditions, rhs[index].Conditions) {
+					return false
+				}
+			}
+		}
+	}
+
+	return true
+}
+
+func (clusterRequest *ClusterKibanaRequest) removeKibana() (err error) {
```

Comment:
I may have missed it... where do we call this from?

Reply:
Fixed by calling that if `clusterRequest == nil` in `ReconcileKibana`

Reply:
so, just a heads up since the `kibana` cr should be the owner of these objects the descheduler will actually remove these objects when the CR is removed... we had it in the CLO before because the `clusterlogging` instance could still be there, but also if the type for visualization wasn't `kibana` we wanted to remove it... so this function may not be necessary anymore